torch/distributed/nn/_functions.py

```diff
@@ -0,0 +1,339 @@
+    """
+    Reduces the tensor data on multiple GPUs across all machines. Each tensor
+    in ``tensor_list`` should reside on a separate GPU
+
+    Only the GPU of ``tensor_list[dst_tensor]`` on the process with rank ``dst``
+    is going to receive the final result.
+
+    Only nccl backend is currently supported
+    tensors should only be GPU tensors
+
+    Arguments:
+        tensor_list (List[Tensor]): Input GPU tensors of the collective.
+            You also need to make sure that ``len(tensor_list)`` is the same for
+            all the distributed processes calling this function.
+        dst (int): Destination rank
+        op (optional): One of the values from
+            ``torch.distributed.ReduceOp``
+            enum.  Specifies an operation used for element-wise reductions.
+        group (ProcessGroup, optional): The process group to work on
+        dst_tensor (int, optional): Destination tensor rank within
+                                    ``tensor_list``
+
+    Returns:
+        list[Tensor]: received tensors from the collective
+
+    """
+    return _ReduceMultiGPU.apply(dst, op, group, dst_tensor, *tensors)
+
+
+def all_gather(tensor, group=dist.group.WORLD):
+    """
+    Gathers tensors from the whole group in a list.
+
+    Arguments:
+        tensor (Tensor): Tensor to be broadcast from current process.
+        group (ProcessGroup, optional): The process group to work on
+
+    Returns:
+        list[Tensor]): output of the collective
+
+    """    
+    return _AllGather.apply(group, tensor)
+
+
+def all_to_all(tensors, group=dist.group.WORLD):
+    """
+    Each process scatters list of input tensors to all processes in a group and
+    return gathered list of tensors in output list.
+
+    Arguments:
+        tensors (list[Tensor]): List of tensors to scatter one per rank.
+        group (ProcessGroup, optional): The process group to work on.
+
+    Returns:
+        list[Tensor]): output of the collective
+
+    """
+    return _AlltoAll.apply(group, *tensors)
+
+
+def all_reduce(tensor, op=dist.ReduceOp.SUM, group=dist.group.WORLD):
+    """
+    Reduces the tensor data across all machines in such a way that all get
+    the final result.
+
+    After the call ``tensor`` is going to be bitwise identical in all processes.
+
+    Arguments:
+        tensor (Tensor): Input of the collective.
+        op (optional): One of the values from
+            ``torch.distributed.ReduceOp``
+            enum.  Specifies an operation used for element-wise reductions.
+        group (ProcessGroup, optional): The process group to work on
+
+    Returns:
+        Tensor: output of the collective
+
+    """
+    return _AllReduce.apply(op, group, tensor)
+
+
+class _Broadcast(Function):
+    @staticmethod
+    def forward(ctx, src, group, tensor):
+        ctx.src = src
+        ctx.group = group
+        ctx.rank = dist.get_rank()
+        dist.broadcast(tensor, src, group=group)
```

Comment:
Is `tensor` modified inplace here?
When you do this, the result of the function is a view of the input. This has some important implications if a user tries to change the output inplace. In this particular case, you will get a warning and the backward function won't be called. I hope to change that to an error in the future but it is still only deprecated for now.

Reply:
yes, tensor is modified in place, actually I verified it with a script and no warning was called, and the backward worked properly  😖

Reply:
You need to use "mark_dirty()" properly for the warning to happen!
When you use custom functions, the automatic inplace detection does not work, you have to do it yourself!

Reply:
as soon as I mark it as dirty nothing works, I decided to clone and detach the input tensor on the `torch.distributed` operations that modify it inplace. Is this approach ok?

Reply:
No need to `.detach()` in the forward as it runs in no_grad mode anyway. But yes cloning the input is good!